compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -548,24 +554,41 @@ trait ConstraintHandling {
         wideInst.dropRepeatedAnnot
   end widenInferred
 
+  /** Convert all toplevel union types in `tp` to hard unions */
+  extension (tp: Type) private def hardenUnions(using Context): Type = tp.widen match
```

Comment:
My original idea was to change https://github.com/lampepfl/dotty/blob/6540ad9154797164561281080b6a4e24a191691d/compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala#L266 which currently uses `|` (and thus creates soft unions) to instead check a flag to decide whether to use a soft or hard union. This would avoid having to convert unions after creating them, and it means we can reset the flag after adding the constraint.

Reply:
The problem I see with this is that the lower bound might be a union with type variables in it. In that case the type variables would end up in the ordering constraint and there would be no union to harden until the time we instantiate the type variable.

Reply:
I see, so we couldn't reset the flag but could check `hardVars` in `ConstraintHandling` when we create the unions I think, that would cover unions created later from propagating constraints based on ordering.

Reply:
I think `hardVars` belongs in `TyperState`. If it's in `ConstraintHandling` we would lose it when we commit a constraint.

Reply:
Right, I meant we could leave it in TyperState but check it in ConstraintHandling#addOneBound instead of checking it in TypeVar#instantiate, but I don't know if we're actually supposed to access the currrent TyperState from ConstraintHandling.

Reply:
But `addOneBound` does not necessarily create a union. The union might be created by fullBounds when we instantiate the variable.

Reply:
But as far as I can tell we don't use `fullBounds` to instantiate a type variable (except for GADTs but we know that's problematic: https://github.com/lampepfl/dotty/pull/14288)

Reply:
We do use `fullLowerBound` in `approximation`.

Reply:
Ah that's right, so we'd have to check hardVars both in addOneBound and fullLowerBound, I'm neutral on whether this is better than the current scheme.

Reply:
I think it's better to tie it to typevariable instantiation, since we do `fullLowerBound` also elsewhere. We could possibly add it to `approximate`, but maybe that mixes up things too much.